### DISCUSSION TITLE
Dynamic field / Need help modeling a document

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#14124] [General]
Hi,

I'm having a problem identifying a proper way to model a document.

Let's assume we have `Vehicle Operations` which describe
operations related to a vehicle like `FuelAdd`, `Repair`, `TireChange`, `CarWash` etc...

Each of these operations has specific metadata.

The metadata could be a dictionary where the key would be a
metadata type like `fuelVolume` with the corresponding value
but the tricky part is that
the value may not always be a primitive but may have more data
that belongs together _(like the RepairMetadata.Part below where 
quantity and comment relate to a specific part)_ 

```csharp
public enum OperationType { FuelAdd, Repair, CarWash, TireChange, }

public class VehicleOperation
{
    ...
    public OperationType Type {get;}
    public object Metadata {get;}
}

public class FuelMetadata {
    public double Volume {get;}
    public string Unit {get;}
    public Money Price {get;}
}

public class RepairMetadata {
    public Part[] Parts {get;}
    
    public class Part {
        public string Id {get;}
        public int Qty {get;}
        public string Comment {get;}
    }
}
```

When retrieving data from RavenDb, one option would be to serialize
the metadata object and deserialize it into the expecting metadata type (doesn't sound performance friendly).
But in case I need to do some indexing, I guess it'll get messy easily?

Any suggestions what would be the RavenDb way to model this? :)

**Comment by ayende** [comment]
Why aren't you using polymorphism in this case?

```
public class VehicleOperation
{

}

public class FuelVehicleOperation : VehicleOperation {
    public double Volume {get;}
    public string Unit {get;}
    public Money Price {get;}
}

public class RepairVehicleOperation  : VehicleOperation {
    public Part[] Parts {get;}
    
    public class Part {
        public string Id {get;}
        public int Qty {get;}
        public string Comment {get;}
    }
}
```


So each "metadata" is just plain data on the `VehicleOperation`. And in your code, you can just accept `VehicleOperation` and use it.
If you need access to the internal data, you cast to the right type or use virtual methods, etc.

**Reply by mcrio** [reply]
Thanks Oren,

> Why aren't you using polymorphism in this case?

I can't recall why I gave up the polymorphism way. Will look into it again. Too much coffee I guess :)
